pkg/pillar/hypervisor/kubevirt.go

```diff
@@ -0,0 +1,1421 @@
+		return logError("can't fetch the PCI device info from kubevirt %v", err)
+	}
+
+	// Get the currently registered  devices from Kubevirt
+	pciHostDevs := kubeVirt.Spec.Configuration.PermittedHostDevices.PciHostDevices
+
+	for i, pa := range pciAssignments {
+
+		vendor, err := pa.vid()
+		if err != nil {
+			return logError("can't fetch the vendor id for pci device %v", err)
+		}
+		// Delete 0x prefix it exists, kubevirt does not like it
+		if strings.HasPrefix(vendor, "0x") {
+			vendor = vendor[2:]
+		}
+		devid, err := pa.devid()
+		if err != nil {
+			return logError("can't fetch the device id for pci device %v", err)
+		}
+		if strings.HasPrefix(devid, "0x") {
+			devid = devid[2:]
+		}
+		pciVendorSelector := vendor + ":" + devid
+
+		// Check if we already registered this device with kubevirt. If not register with kubevirt
+		registered := isRegisteredPciHostDevice(pciVendorSelector, pciHostDevs)
+		// Lets make sure resname is unique by appending vendor and devid
+		// NOTE we cannot use pciVendorSelector directly since ":" is not accepted in kubernetes resource name standard
+		resname := "devices.kubevirt.io/hostdevice-" + vendor + devid
+		if !registered {
+
+			newpcidev := v1.PciHostDevice{
+				ResourceName:      resname,
+				PCIVendorSelector: pciVendorSelector,
+			}
+			logrus.Infof("Registering PCI device %s as resource %s with kubevirt", pciVendorSelector, resname)
+			kubeVirt.Spec.Configuration.PermittedHostDevices.PciHostDevices = append(kubeVirt.Spec.Configuration.PermittedHostDevices.PciHostDevices, newpcidev)
+			_, err = kv.Update(kubeVirt)
+
+			if err != nil {
+				return logError("can't update the PCI device info from kubevirt %v", err)
+			}
+		}
+		// At this point we have registered the PCI device with kubevirt
+		// Create HostDevice array which will be inserted into vmi Hostdevices
+		// Hostdevices could be NVMe drives,USB or NICs, that is reason if just call them device.
+
+		pcidevices[i] = v1.HostDevice{
+			DeviceName: resname,
+			Name:       "device" + strconv.Itoa(i+1),
+		}
+
+		vmi.Spec.Domain.Devices.HostDevices = append(vmi.Spec.Domain.Devices.HostDevices, pcidevices[i])
+	}
+
+	return nil
+
+}
+
+func isRegisteredPciHostDevice(pciVendorSelector string, PciHostDevices []v1.PciHostDevice) bool {
+
+	for _, dev := range PciHostDevices {
+
+		if dev.PCIVendorSelector == pciVendorSelector {
+			return true
+		}
+	}
+
+	return false
+}
+
+func addKernelBootContainer(spec *v1.VirtualMachineInstanceSpec, image, kernelArgs, kernelPath, initrdPath string) *v1.VirtualMachineInstanceSpec {
+	if spec.Domain.Firmware == nil {
+		spec.Domain.Firmware = &v1.Firmware{}
+	}
+
+	spec.Domain.Firmware.KernelBoot = &v1.KernelBoot{
+		KernelArgs: kernelArgs,
+		Container: &v1.KernelBootContainer{
+			Image:           image,
+			KernelPath:      kernelPath,
+			InitrdPath:      initrdPath,
+			ImagePullPolicy: k8sv1.PullNever,
+		},
+	}
+
+	return spec
+}
+
+// PCIReserve reserves a PCI device for Kubevirt
```

Comment:
Its exactly same. But under different hypervisor. We tried to use directly from kvm.go by not defining in kubevirt.go. That did not work, for some reason its randomly invoking PCIReserve/PCIRelease from other hypervisors. So decided to copy those explicitly. Do you know of a better way, let me know.

Reply:
Perhaps you created an instance of `xenContext` and called `PCIReserve` on that one?

I guess it would be best, to move the shared code between `KvmContext` and kubevirt into a new file and use funcs without receivers.

Reply:
Then perhaps create a base class and so that both can embedd it, that makes it clear to the reader that kvm and kubevirt are very similar in this regard.

Reply:
I moved those to common functions under hypervisor.go and gets called from both kvm.go and Kubevirt.go. You will see in next commit.